File: MorkosiaPrepaLeague/Models/Core/League.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;

namespace MorkosiaPrepaLeague.Models.Core
{
    public class League : CosmosDocumentBase
    {
        [Required]
        [Range(3, 50, ErrorMessage = "League name length must be between 1 and 50")]
        public string Name { get; set; } = default!;

        [Required]
        [RegularExpression("^(ActivityPoints|H2H)$", ErrorMessage = "Type must be 'ActivityPoints' or 'H2H'")]
        public string Type { get; set; } = default!;

        [Range(0, int.MaxValue, ErrorMessage = "Initial budget must be a non-negative value")]
        public int? InitialBudget { get; set; }

        /// <summary>
        /// Team structure: Collection of positions and how many players are needed for each position
        /// </summary>
        public List<TeamPosition>? TeamPositions { get; set; }

        [Range(0, int.MaxValue, ErrorMessage = "Default player price must be a non-negative value")]
        public int? DefaultPlayerPrice { get; set; }

        /// <summary>
        /// Player-specific price overrides: Key = UserId, Value = price for that player in this league.
        /// </summary>
        public Dictionary<string, int>? PlayerPriceOverrides { get; set; }

        /// <summary>
        /// Team/Manager-specific budget overrides: Key = ManagerUserId, Value = custom budget for that manager's team in this league.
        /// If a manager is not in this dictionary, they use the InitialBudget value.
        /// </summary>
        public Dictionary<string, int>? BudgetOverrides { get; set; }

        /// <summary>
        /// Default ownership cap for any member-position combination that doesn't have a specific override.
        /// If null, no default cap is enforced.
        /// </summary>
        [Range(0, int.MaxValue, ErrorMessage = "Default ownership cap must be a non-negative value")]
        public int? DefaultOwnershipCap { get; set; }

        /// <summary>
        /// Member-specific position-based ownership caps. 
        /// Key = MemberUserId, Value = Dictionary where Key = Position name, Value = Max teams for that member-position combination.
        /// Example: { "user123": { "GK": 1, "DEF": 2 }, "user456": { "MID": 3 } }
        /// </summary>
        public Dictionary<string, Dictionary<string, int>>? MemberPositionOwnershipCaps { get; set; }

        /// <summary>
        /// Calculates the total team size based on the sum of all position counts.
        /// </summary>
        /// <returns>The total team size or null if positions are not defined.</returns>
        public int? GetTotalTeamSize()
        {
            if (TeamPositions != null && TeamPositions.Any())
            {
                return TeamPositions.Sum(p => p.Count);
            }
            
            return null;
        }

        /// <summary>
        /// Gets the effective budget for a specific manager, considering budget overrides.
        /// </summary>
        /// <param name="managerUserId">The user ID of the manager</param>
        /// <returns>The budget amount for the manager, or null if no budget is configured</returns>
        public int? GetEffectiveBudget(string managerUserId)
        {
            if (string.IsNullOrEmpty(managerUserId))
                return InitialBudget;

            // Check if there's a specific budget override for this manager
            if (BudgetOverrides?.ContainsKey(managerUserId) == true)
                return BudgetOverrides[managerUserId];

            // Fall back to the initial budget
            return InitialBudget;
        }

        /// <summary>
        /// Gets the effective ownership cap for a specific member-position combination.
        /// </summary>
        /// <param name="memberUserId">The user ID of the member</param>
        /// <param name="positionName">The name of the position</param>
        /// <returns>The ownership cap for the member-position combination, or null if no cap is configured</returns>
        public int? GetEffectiveOwnershipCap(string memberUserId, string positionName)
        {
            if (string.IsNullOrEmpty(memberUserId) || string.IsNullOrEmpty(positionName))
                return DefaultOwnershipCap;

            // Check if there's a specific cap for this member-position combination
            if (MemberPositionOwnershipCaps?.ContainsKey(memberUserId) == true &&
                MemberPositionOwnershipCaps[memberUserId]?.ContainsKey(positionName) == true)
            {
                return MemberPositionOwnershipCaps[memberUserId][positionName];
            }

            // Fall back to the default ownership cap
            return DefaultOwnershipCap;
        }

        /// <summary>
        /// Gets the price for a specific player in this league, considering overrides and defaults.
        /// </summary>
        /// <param name="playerId">The player's user ID</param>
        /// <returns>The price for the player</returns>
        public int GetPlayerPrice(string playerId)
        {
            // Check if there's a specific price override for this player
            if (PlayerPriceOverrides?.ContainsKey(playerId) == true)
            {
                return PlayerPriceOverrides[playerId];
            }

            // Fall back to the default player price, or 0 if not set
            return DefaultPlayerPrice ?? 0;
        }
    }

    public class TeamPosition
    {
        [Required]
        public string Name { get; set; } = string.Empty;
        
        [Required]
        [Range(1, int.MaxValue, ErrorMessage = "Count must be at least 1")]
        public int Count { get; set; }
    }
}

File: src/app/models/league.py
from typing import Dict, List, Optional

from pydantic import BaseModel, Field

from app.models.base import DocumentBase


class TeamPosition(BaseModel):
    """A single position slot in a team (e.g. name='GK', count=1)."""
    name: str
    count: int = Field(ge=1)


class League(DocumentBase):
    name: str
    type: str  # "ActivityPoints" or "H2H"
    initial_budget: Optional[int] = None
    team_positions: Optional[List[TeamPosition]] = None
    default_player_price: Optional[int] = None

    # Key: player user_id → custom price
    player_price_overrides: Optional[Dict[str, int]] = None

    # Key: manager user_id → custom budget
    budget_overrides: Optional[Dict[str, int]] = None

    # Cap applied when no member-specific cap exists for a position
    default_ownership_cap: Optional[int] = None

    # Key: member user_id → { position_name: max_teams }
    member_position_ownership_caps: Optional[Dict[str, Dict[str, int]]] = None

    class Settings:
        name = "leagues"

    # ------------------------------------------------------------------
    # Business logic helpers (ported directly from League.cs methods)
    # ------------------------------------------------------------------

    def get_total_team_size(self) -> Optional[int]:
        if self.team_positions:
            return sum(p.count for p in self.team_positions)
        return None

    def get_effective_budget(self, manager_user_id: str) -> Optional[int]:
        if self.budget_overrides and manager_user_id in self.budget_overrides:
            return self.budget_overrides[manager_user_id]
        return self.initial_budget

    def get_effective_ownership_cap(
        self, member_user_id: str, position_name: str
    ) -> Optional[int]:
        if self.member_position_ownership_caps:
            member_caps = self.member_position_ownership_caps.get(member_user_id)
            if member_caps and position_name in member_caps:
                return member_caps[position_name]
        return self.default_ownership_cap

    def get_player_price(self, player_id: str) -> int:
        if self.player_price_overrides and player_id in self.player_price_overrides:
            return self.player_price_overrides[player_id]
        return self.default_player_price or 0
